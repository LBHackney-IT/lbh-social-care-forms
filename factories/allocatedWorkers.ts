import { Factory } from 'fishery';

import { Allocation } from 'types';

export const allocationFactory = Factory.define<Allocation>(({ sequence }) => ({
  id: sequence,
  allocatedWorkerTeam: 'Safeguarding and Reviewing Team',
  allocatedWorker: 'Officer Name',
  allocationEndDate: '2019-03-28 00:00:00',
  allocationStartDate: '2019-03-28 00:00:00',
  workerType: 'Consultant Social Worker',
  personId: 1234,
  caseStatus: 'Open',
  personName: 'foo',
  personAddress: 'the address',
  personDateOfBirth: '2020-03-20',
  ragRating: 'amber',
}));

export const mockedAllocation = allocationFactory.build();
export const mockedAllocations = [allocationFactory.build()];
