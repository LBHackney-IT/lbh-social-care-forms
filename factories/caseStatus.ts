import { Factory } from 'fishery';
import {
  CaseStatus,
  CaseStatusFields,
  AddCaseStatusFormData,
  CaseStatusAnswerDisplay,
} from 'types';

export const mockedCaseStatusFactory = Factory.define<CaseStatus>(
  ({ sequence }) => ({
    id: sequence,
    type: 'CIN',
    answers: [mockedStatusField.build()],
    startDate: '2021-01-01T02:00:00Z',
    endDate: '2021-12-01T02:00:00Z',
    notes: 'a lot to note',
  })
);

export const mockedCaseStatusAddRequest = Factory.define<AddCaseStatusFormData>(
  ({ sequence }) => ({
    personId: sequence,
    type: 'CIN',
    fields: [{ name: 'test', selected: 'test' }],
    startDate: '2021-01-01T02:00:00Z',
    endDate: '2021-12-01T02:00:00Z',
    notes: 'a lot to note',
    createdby: 'test@testington.com',
  })
);

export const mockedStatusField = Factory.define<CaseStatusFields>(() => ({
  option: 'placementType',
  value: 'K1',
  startDate: '2020-08-01',
  createdAt: '2021-10-11T10:54:32Z',
}));

export const mockedCaseStatusAnswers = Factory.define<CaseStatusAnswerDisplay>(
  () => ({
    startDate: '2021-09-10',
    endDate: '2021-10-10',
    status: [mockedStatusField.build()],
  })
);
