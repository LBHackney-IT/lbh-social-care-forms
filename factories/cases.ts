import { Factory } from 'fishery';

import { Case } from 'types';

export const caseFactory = Factory.define<Case>(({ sequence }) => ({
  recordId: sequence.toString(),
  personId: 123,
  dateOfEvent: '25/10/2020 13:49:43',
  formName: 'foorm',
  officerEmail: 'Fname.Lname@hackney.gov.uk',
  caseFormData: {
    note: 'I am the note',
    first_name: 'Foo',
    last_name: 'Bar',
    timestamp: '11/03/2021 16:51:56',
    mosaic_id: 123,
    form_name_overall: 'foo',
    worker_email: 'worker@hackney.gov.uk',
    form_name: 'Foo bar',
    context_flag: 'A',
    date_of_event: '25/10/2020 13:49:43',
  },
  deleted: false,
  isImported: false,
}));

export const mockedNote = caseFactory.build({ caseFormUrl: 'https://foo.bar' });
export const mockedAllocationNote = caseFactory.build({
  formName: 'Worker allocated',
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    form_name_overall: 'API_Allocation',
    form_name: 'Worker allocated',
  },
});
export const mockedDeallocationNote = caseFactory.build({
  formName: 'Worker allocated',
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    deallocation_reason: 'deallocated because',
    form_name_overall: 'API_Deallocation',
    form_name: 'Worker deallocated',
  },
});
export const mockedCaseNote = caseFactory.build({
  caseFormData: {
    form_name_overall: 'ASC_case_note',
    case_note_title: 'i am a case title',
  },
});
export const mockedDeletedCaseNote = caseFactory.build({
  caseFormData: {
    form_name_overall: 'ASC_case_note',
    case_note_title: 'i am a case title',
  },
  formName: 'adult-case-note',
  deleted: true,
});
export const mockedWarningNoteCase = caseFactory.build({
  formName: 'Warning Note',
  caseFormData: {
    form_name_overall: 'API_WarningNote',
    warning_note_id: 456,
  },
});
