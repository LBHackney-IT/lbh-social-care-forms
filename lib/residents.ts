import axios from 'axios';

import type { LegacyResident, Resident, ResidentsAPI, User } from 'types';
import { getAuditingParams } from '../utils/auditing';

const ENDPOINT_API = process.env.ENDPOINT_API;
const AWS_KEY = process.env.AWS_KEY;

const headers = { 'x-api-key': AWS_KEY };

interface ResidentBE extends LegacyResident {
  addressList: Array<{
    displayAddressFlag: 'Y' | 'N';
    addressLine1: string;
    postCode: string;
    uprn: string;
  }>;
}

const sanitiseResidentData = (residents: ResidentBE[]): LegacyResident[] =>
  residents?.map(({ addressList, ...resident }: ResidentBE): LegacyResident => {
    const address = addressList?.find(
      ({ displayAddressFlag }) => displayAddressFlag === 'Y'
    );
    return {
      ...resident,
      address: address && {
        address: address.addressLine1,
        postcode: address.postCode,
        uprn: address.uprn,
      },
    };
  });

export const getResidents = async (
  params: Record<string, unknown>
): Promise<ResidentsAPI> => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents`, {
    headers,
    params,
  });
  return { ...data, residents: sanitiseResidentData(data.residents) };
};

export const getResident = async (
  personId: number,
  user: User
): Promise<Resident> => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents/${personId}`, {
    headers,
    params: getAuditingParams(user),
  });
  return data;
};

export const normalisePhoneInput = (
  formData: Resident
): Record<string, unknown> => ({
  ...formData,
  phoneNumbers: formData.phoneNumbers
    ? formData.phoneNumbers.map((phone) => ({
        ...phone,
        type: phone.type === '' ? 'main' : phone.type,
      }))
    : null,
});

export const addResident = async (
  formData: Resident
): Promise<ResidentsAPI> => {
  const { data } = await axios.post(
    `${ENDPOINT_API}/residents`,
    normalisePhoneInput(formData),
    {
      headers: { ...headers, 'Content-Type': 'application/json' },
    }
  );
  return data;
};

export const updateResident = async (formData: Resident): Promise<Resident> => {
  await axios.patch(
    `${ENDPOINT_API}/residents`,
    normalisePhoneInput(formData),
    {
      headers: { ...headers, 'Content-Type': 'application/json' },
    }
  );
  return formData;
};
