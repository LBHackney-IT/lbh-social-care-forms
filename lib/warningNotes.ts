import axios from 'axios';

import { WarningNote } from 'types';

const ENDPOINT_API = process.env.ENDPOINT_API;
const AWS_KEY = process.env.AWS_KEY;

const headers = { 'x-api-key': AWS_KEY };

export const getWarningNotesByResident = async (
  personId: number
): Promise<WarningNote[] | []> => {
  const { data }: { data: { warningNotes: WarningNote[] } } = await axios.get(
    `${ENDPOINT_API}/residents/${personId}/warningnotes`,
    {
      headers,
    }
  );
  return data.warningNotes.filter((note) => note.status === 'open');
};

export const getWarningNote = async (
  warningNoteId: number
): Promise<WarningNote | undefined> => {
  const { data } = await axios.get(
    `${ENDPOINT_API}/warningnotes/${warningNoteId}`,
    {
      headers,
    }
  );
  return data;
};

export const addWarningNote = async (
  params: Record<string, unknown>
): Promise<void> => {
  await axios.post(`${ENDPOINT_API}/warningnotes`, params, {
    headers: { ...headers, 'Content-Type': 'application/json' },
  });
};

export const updateWarningNote = async (
  params: Record<string, unknown>
): Promise<void> => {
  await axios.patch(`${ENDPOINT_API}/warningnotes`, params, {
    headers: { ...headers, 'Content-Type': 'application/json' },
  });
};
