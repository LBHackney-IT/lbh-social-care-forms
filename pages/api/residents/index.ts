import { StatusCodes } from 'http-status-codes';

import { getResidents, addResident } from 'lib/residents';
import { isAuthorised } from 'utils/auth';
import { apiHandler } from 'lib/apiHandler';
import { middleware as csrfMiddleware } from 'lib/csrfToken';

import type { NextApiRequest, NextApiResponse, NextApiHandler } from 'next';
import { AxiosError } from 'axios';

const endpoint: NextApiHandler = async (
  req: NextApiRequest,
  res: NextApiResponse
) => {
  const user = isAuthorised(req);
  if (!user) {
    res.status(StatusCodes.UNAUTHORIZED);
    return;
  }
  if (!user.isAuthorised) {
    res.status(StatusCodes.FORBIDDEN);
    return;
  }
  switch (req.method) {
    case 'GET':
      try {
        const data = await getResidents(req.query);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error(
          'Residents get error:',
          (error as AxiosError)?.response?.data
        );
        (error as AxiosError)?.response?.status === StatusCodes.NOT_FOUND
          ? res
              .status(StatusCodes.NOT_FOUND)
              .json({ message: 'Residents Not Found' })
          : res
              .status(StatusCodes.INTERNAL_SERVER_ERROR)
              .json({ message: 'Unable to get the Residents' });
      }
      break;

    case 'POST':
      try {
        const data = await addResident(req.body);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error(
          'Resident post error:',
          (error as AxiosError)?.response?.data
        );
        console.error('Resident post request:', req);
        res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Unable to add resident' });
      }
      break;

    default:
      res
        .status(StatusCodes.BAD_REQUEST)
        .json({ message: 'Invalid request method' });
  }
};

export default apiHandler(csrfMiddleware(endpoint));
