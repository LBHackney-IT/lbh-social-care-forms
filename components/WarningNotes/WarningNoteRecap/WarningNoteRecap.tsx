import Spinner from 'components/Spinner/Spinner';
import Summary from 'components/Summary/Summary';
import ErrorMessage from 'components/ErrorMessage/ErrorMessage';
import { useWarningNote } from 'utils/api/warningNotes';
import { formStepsAdult, formStepsChild } from 'data/forms/warning-note';
import { Resident } from 'types';
import SummaryList from '../../Summary/SummaryList';
import {
  reviewFormStepsAdult,
  reviewFormStepsChild,
} from 'data/forms/warning-note-review';

export interface Props {
  person: Resident;
  warningNoteId: number;
  noteDetails?: string;
}

const WarningNoteRecap = ({
  person,
  warningNoteId,
  noteDetails,
}: Props): React.ReactElement => {
  const { data: warningNote, error: warningNoteError } =
    useWarningNote(warningNoteId);

  if (warningNoteError) {
    return <ErrorMessage />;
  }
  if (!warningNote) {
    return <Spinner />;
  }

  const shouldPageDisplayNoteReviews = (
    noteDetails: string | undefined
  ): boolean => {
    return noteDetails !== undefined && noteDetails !== 'note-created';
  };

  const reviewsOrderedByMostRecentFirst = [
    ...warningNote.warningNoteReviews,
  ].reverse();

  const reviewFormSteps =
    person.contextFlag === 'A' ? reviewFormStepsAdult : reviewFormStepsChild;

  return (
    <>
      {shouldPageDisplayNoteReviews(noteDetails) &&
        reviewsOrderedByMostRecentFirst.map((review) => {
          const isSelectedReviewTheMostRecent =
            reviewsOrderedByMostRecentFirst.indexOf(review) === 0;
          return (
            <Summary
              key={review.id}
              formData={{
                ...review,
                disclosedWithIndividual: review.disclosedWithIndividual
                  ? 'Yes'
                  : 'No',
                outputAsDetailedSummary: 'Yes',
                endDate: isSelectedReviewTheMostRecent
                  ? warningNote.endDate
                  : null,
                nextReviewDate: isSelectedReviewTheMostRecent
                  ? warningNote.nextReviewDate
                  : null,
              }}
              formSteps={reviewFormSteps.map((step) => {
                return {
                  ...step,
                  title: `Warning Review Details ${new Date(
                    review.reviewDate
                  ).toLocaleDateString('en-GB')}`,
                };
              })}
              formPath="/people/:peopleId/warning-notes/:warningNoteId"
            />
          );
        })}
      <Summary
        formData={{
          ...warningNote,
          disclosedWithIndividual: warningNote.disclosedWithIndividual
            ? 'Yes'
            : 'No',
        }}
        formSteps={person.contextFlag === 'A' ? formStepsAdult : formStepsChild}
        formPath={`/people/:peopleId/warning-notes/:warningNoteId`}
        additionalMetadata={[
          {
            key: 'created_by',
            title: 'Created by',
            value: warningNote.createdBy,
          },
        ]}
      />
    </>
  );
};

export default WarningNoteRecap;
