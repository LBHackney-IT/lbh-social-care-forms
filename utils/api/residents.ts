import axios from 'axios';
import useSWR, { useSWRInfinite, SWRResponse, SWRInfiniteResponse } from 'swr';

import { getInfiniteKey } from 'utils/api';

import type { Resident, ErrorAPI } from 'types';

export const useResidents = (
  params: Record<string, unknown>,
  invoke = true
): SWRInfiniteResponse<Record<string, unknown>, Error> =>
  useSWRInfinite(
    // @ts-ignore
    invoke ? getInfiniteKey('/api/residents', 'residents', params) : null
  );

export const useResident = (
  personId: number
): SWRResponse<Resident, ErrorAPI> => useSWR(`/api/residents/${personId}`);

export const addResident = async (
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.post(`/api/residents`, formData);
  return { ref: data?.id, data };
};

export const updateResident = async (
  personId: number,
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.patch(`/api/residents/${personId}`, formData);
  return { ref: data?.id, data };
};

export const patchResident = async (formData: {
  id: number;
  createdBy: string;
  [key: string]: unknown;
}): Promise<Record<string, unknown>> => {
  const { data } = await axios.patch(`/api/residents`, formData);
  return { ref: data?.id, data };
};
