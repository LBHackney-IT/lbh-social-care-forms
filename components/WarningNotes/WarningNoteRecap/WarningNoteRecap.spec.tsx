import { render } from '@testing-library/react';

import WarningNoteRecap from './WarningNoteRecap';
import * as warningNotes from 'utils/api/warningNotes';
import { mockedWarningNote, warningNoteFactory } from 'factories/warningNotes';
import { mockedResident } from 'factories/residents';

jest.mock('components/Spinner/Spinner', () => () => 'MockedSpinner');

describe(`WarningNoteRecap`, () => {
  const props = {
    person: mockedResident,
    warningNoteId: 1,
  };

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should render correctly', async () => {
    jest.spyOn(warningNotes, 'useWarningNote').mockImplementation(() => ({
      data: mockedWarningNote.find(({ id }) => id === props.warningNoteId),
      mutate: jest.fn(),
      revalidate: jest.fn(),
      isValidating: false,
    }));

    const { asFragment } = render(<WarningNoteRecap {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should show the email address of the user who created the warning note', async () => {
    jest.spyOn(warningNotes, 'useWarningNote').mockImplementation(() => ({
      data: warningNoteFactory.build({
        createdBy: 'first.last@hackney.gov.uk',
      }),
      mutate: jest.fn(),
      revalidate: jest.fn(),
      isValidating: false,
    }));

    const { getByText } = render(
      <WarningNoteRecap person={mockedResident} warningNoteId={0} />
    );

    getByText('first.last@hackney.gov.uk');
  });
});
