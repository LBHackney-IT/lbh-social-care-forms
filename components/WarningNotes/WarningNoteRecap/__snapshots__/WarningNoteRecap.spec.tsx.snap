// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WarningNoteRecap should render correctly 1`] = `
<DocumentFragment>
  <div
    class="govuk-!-margin-bottom-7"
  >
    <div
      class="lbh-table-header"
    >
      <div
        class="sectionTitle"
      >
        <h3
          class="govuk-heading-m govuk-custom-text-color govuk-!-margin-bottom-0"
        >
          WARNING DETAILS
        </h3>
      </div>
    </div>
    <hr
      class="govuk-divider"
    />
    <dl
      class="govuk-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Warning type
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Risk to Adults
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Start date
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          22/11/2020
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Review / end date
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          01/12/2020
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Is the individual aware of the Warning Note and Review Date?
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Yes
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Date informed
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          12/11/2020
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          How was the individual informed
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Written
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Details of disclosure to individual
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          <div
            class="outer"
          >
            <p
              data-testid="para"
            >
              lorem ipsum
            </p>
          </div>
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Warning narrative and risks notes
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          <div
            class="outer"
          >
            <p
              data-testid="para"
            >
              i am a note
            </p>
          </div>
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Managerâ€™s name
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          the big boss
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Date discussed with manager
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          12/11/2020
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Created by
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Foo
        </dd>
      </div>
    </dl>
  </div>
</DocumentFragment>
`;
